clippy_lints/src/avoidable_slice_indexing.rs

```diff
@@ -0,0 +1,291 @@
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::mir::FakeReadCause;
+use rustc_middle::ty;
+use rustc_semver::RustcVersion;
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::{symbol::Ident, Span, Symbol};
+use rustc_trait_selection::infer::TyCtxtInferExt;
+use rustc_typeck::expr_use_visitor::{Delegate, ExprUseVisitor, Place, PlaceBase, PlaceWithHirId};
+use std::convert::TryInto;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// The lint checks for slices that come from value deconstruction and
+    /// are only used to access individual slice values.
+    ///
+    /// ### Why is this bad?
+    /// Accessing slice values using indices can lead to panics.
+    ///
+    /// ### Limitations
+    /// This lint currently only checks for immutable access.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let slice: Option<&[u32]> = Some(&[1, 2, 3]);
+    ///
+    /// if let Some(slice) = slice {
+    ///     println!("{}", slice[0]);
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let slice: Option<&[u32]> = Some(&[1, 2, 3]);
+    ///
+    /// if let Some(&[first, ..]) = slice {
+    ///     println!("{}", first);
+    /// }
+    /// ```
+    pub AVOIDABLE_SLICE_INDEXING,
+    nursery,
+    "avoid indexing on slices which could be deconstructed"
+}
+
+#[derive(Copy, Clone)]
+pub struct AvoidableSliceIndexing {
+    indexing_limit: u64,
+    msrv: Option<RustcVersion>,
+}
+
+impl AvoidableSliceIndexing {
+    pub fn new(avoidable_slice_indexing_limit: u64, msrv: Option<RustcVersion>) -> Self {
+        Self {
+            indexing_limit: avoidable_slice_indexing_limit,
+            msrv,
+        }
+    }
+}
+
+impl_lint_pass!(AvoidableSliceIndexing =>  [AVOIDABLE_SLICE_INDEXING]);
+
+impl LateLintPass<'_> for AvoidableSliceIndexing {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'_>) {
+        if_chain! {
+            if !in_macro(expr.span) || is_expn_of(expr.span, "if_chain").is_some();
+            if let Some(IfLet {let_pat, if_then, ..}) = IfLet::hir(cx, expr);
+            if !is_lint_allowed(cx, AVOIDABLE_SLICE_INDEXING, expr.hir_id);
+            if meets_msrv(self.msrv.as_ref(), &msrvs::SLICE_PATTERNS);
+            if let Some(found_slices) = find_slice_values(cx, let_pat);
+            if let Some(filtered_slices) = filter_lintable_slices(cx, found_slices, self.indexing_limit, if_then);
+            then {
+                for slice in filtered_slices.values() {
+                    lint_slice(cx, slice);
+                }
+            }
+        }
+    }
+
+    extract_msrv_attr!(LateContext);
+}
+
+fn find_slice_values(cx: &LateContext<'_>, pat: &hir::Pat<'_>) -> Option<FxHashMap<hir::HirId, SliceLintInformation>> {
+    let mut removed_pat: FxHashSet<Symbol> = FxHashSet::default();
+    let mut ident_to_id: FxHashMap<Symbol, hir::HirId> = FxHashMap::default();
```

Comment:
I like to track bindings by `HirId` rather that `Symbol`. Then I can use `cx.tcx.hir().name(id)` and `cx.tcx.hir().span(id)` when those are needed. Would that remove the need for `ident_to_id` and `pattern_spans`?

Reply:
Each binding inside a pattern has its own `HirId` even if they are later used as the same value. I at first tried to only have the `slices` map in the next line, but that missed other bindings with the same name. We therefore need to match all bindings somehow.

The implementation could use `cx.tcx.hir().name(id)` to avoid `ident_to_id` and use `HirId`s in `removed_pat`. However, each newly found binding would then need to first call `cx.tcx.hir().name(id)` on itself and then on every key inside `ident_to_id` to check if there is already an entry for it. I personally find this more readable and okay, as the scope is limited to only this function. Everything else uses the `HirId` to track the slice. :upside_down_face:

Reply:
> Each binding inside a pattern has its own `HirId` even if they are later used as the same value.

I believe the `Pat` has a unique `HirId` but the `HirId` in `PatKind::Binding` should be the same.

Reply:
Ahh yes, you're correct. I've at first used [`Pat::each_binding()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Pat.html#method.each_binding) which has the described problem. I'll apply this refactoring with the Visitor refactoring :upside_down_face: